File: BFS/127.word-ladder.go
package leetcode

// @leet start
func ladderLength(beginWord string, endWord string, wordList []string) int {
	var (
		queue = []string{beginWord}
		dict  = map[string]bool{}
		step  int
	)
	for _, w := range wordList {
		dict[w] = true
	}
	if !dict[endWord] {
		return 0
	}

	for len(queue) > 0 {
		step++
		for range len(queue) {
			w := queue[0]
			queue = queue[1:]
			if w == endWord {
				return step
			}
			// enqueue all the posibile 1 diff word
			for i := 0; i < len(w); i++ {
				for char := 'a'; char <= 'z'; char++ {
					newW := w[:i] + string(char) + w[i+1:]
					if dict[newW] {
						queue = append(queue, newW)
						dict[newW] = false
					}
				}
			}
		}

	}

	return 0
}

// @leet end

File: BFS/127.word-ladder.py
from collections import deque
from typing import List


# @leet start
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        all_letter = [chr(i + ord("a")) for i in range(26)]

        def possible(s: str):
            for i in range(len(s)):
                for c in all_letter:
                    # NOTE: s[len(s):] will return "" without key error
                    yield s[:i] + c + s[i + 1 :]

        map = set(wordList)
        # if endWord not in map:
        #     return 0
        step = 1
        q = deque([beginWord])
        while len(q) > 0:
            for _ in range(len(q)):
                top = q.popleft()
                if top == endWord:
                    return step
                for var in possible(top):
                    if var in map:
                        q.append(var)
                        map.remove(var)
            step += 1
        return 0


# @leet end

